code/modules/station_goals/bsa.dm

```diff
@@ -305,50 +305,43 @@
 /obj/machinery/computer/bsa_control/attack_hand(mob/user)
 	if(..())
 		return 1
-	ui_interact(user)
+	tgui_interact(user)
 
-/obj/machinery/computer/bsa_control/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+/obj/machinery/computer/bsa_control/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		ui = new(user, src, ui_key, "bsa.tmpl", name, 400, 305)
+		ui = new(user, src, ui_key, "BlueSpaceArtilleryControl", name, 400, 305, master_ui, state)
 		ui.open()
-		ui.set_auto_update(1)
+		ui.set_autoupdate(TRUE)
 
-/obj/machinery/computer/bsa_control/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
+/obj/machinery/computer/bsa_control/tgui_data(mob/user)
 	var/list/data = list()
 	data["connected"] = cannon
 	data["notice"] = notice
 	if(target)
 		data["target"] = get_target_name()
-
 	if(cannon)
 		var/reload_cooldown = cannon.reload_cooldown
 		var/last_fire_time = cannon.last_fire_time
 		var/time_to_wait = max(0, round(reload_cooldown - ((world.time / 10) - last_fire_time)))
 		var/minutes = max(0, round(time_to_wait / 60))
 		var/seconds = max(0, time_to_wait - (60 * minutes))
-
-		data["reloadtime_mins"] = minutes
-		data["reloadtime_secs"] = (seconds < 10) ? "0[seconds]" : seconds
+		var/seconds2 = (seconds < 10) ? "0[seconds]" : seconds
+		data["reloadtime_text"] = "[minutes]:[seconds2]"
 		data["ready"] = minutes == 0 && seconds == 0
 	else
 		data["ready"] = FALSE
-
 	return data
 
-/obj/machinery/computer/bsa_control/Topic(href, href_list)
+/obj/machinery/computer/bsa_control/tgui_act(action, params)
 	if(..())
-		return 1
-
-	if(href_list["build"])
+		return
+	if(action == "build")
 		cannon = deploy()
-		. = TRUE
-	else if(href_list["fire"])
+	if(action == "fire")
 		fire(usr)
-		. = TRUE
-	else if(href_list["recalibrate"])
+	if(action == "recalibrate")
 		calibrate(usr)
-		. = TRUE
```

Comment:
Should return `TRUE` if any of the actions are performed to ensure the UI updates immediately rather than next tick